Add XNNPACK backend option for workspace sharing

GregoryComer opened the pull request:
### Summary
Add a backend option for XNNPACK to enable runtime control of workspace sharing. I've added 3 mode options - Disabled, PerModel, and Global. PerModel shares the workspace between all CALL_DELEGATE instances in a model, keyed by memory allocator address (see below). Global uses a single workspace instance.

I've written the code to allow for the active workspace mode to be safely changed at any time. The workspace instance is resolved at delegate instance init time (model load) and is stored in the XNNExecutor instance. This design will also allow us to set per-model sharing options in the future. I've introduced a wrapper class (XNNWorkspace) to help with synchronization.

With regard to the PerModel behavior, I am using the address of the runtime allocator to disambiguate the model. This is not ideal in the long-run, but there is some larger discussion around generating IDs in a coherent manner in multithreaded environments without synchronization in the core runtime. This might require PAL changes (exposing a thread ID, for example), so I intend to come back to this.

It should be possible to transparently update this logic in the future. The program ID can collide or change without affecting correctness, but may increase memory (for collisions) or enforce extra synchronization (if unstable between delegate instances in a method).

I'd like to add a PerMethod mode as a follow-up. This should be keyed to the specific method instance (not name), such that multiple method instances for the same method can be loaded for execution on different threads without forcing synchronization, but still allow sharing between call delegate instances in each method instance. This will require a unique method identifier.

### Test plan
CI. I've also added a set of dedicated tests for getting/setting the option, running PTEs in each mode, switching modes at runtime, and I've also updated the multithreaded stress test to run in each mode.

digantdesai commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -86,9 +75,29 @@ class XnnpackBackend final
+
+    std::unique_lock<std::mutex> lock(workspace_mutex_, std::defer_lock);
+    if (enable_shared_workspace_) {
+      lock.lock();
```
I guess this will protect us from someone doing `method::init()` on the same program from different threads?

GregoryComer replied:
I've reworked the design since this comment, so marking as resolved.

digantdesai commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -160,9 +170,10 @@ class XnnpackBackend final
       // This is needed to serialize access to xnn_delete_runtime which is not
       // thread safe. This can heppen when multiple threads call destroy() on
```
s/heppen/happen

GregoryComer replied:
Done.

digantdesai commented on backends/xnnpack/runtime/XNNPACKBackend.h:
```diff
@@ -0,0 +1,7 @@
+#include <executorch/runtime/platform/compiler.h>
+
+namespace executorch::backends::xnnpack {
+ET_EXPERIMENTAL void set_workspace_sharing_enabled(bool enable);
```
So from priority, compile_time > runtime? And at runtime, it's one shot at the start up? What about some thread turning it on vs. other off? I guess it's on them?

GregoryComer replied:
I've updated for backend options. To maintain the current behavior, I'm setting global mode if ENABLE_XNNPACK_SHARED_WORKSPACE is set, though we probably want to re-evaluate this in the future, as per-model might be better.

The runtime backend option is global, but I've written the code in such way that we can set it per-method/model if/when backend options support this.

digantdesai reviewed (commented):
Looks good. Please add a test at least to make sure we don't regress on the API and its semantics?

GregoryComer commented:
> backend options is landed btw

Thanks. I'll rework to use those.

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -129,9 +138,10 @@ class XnnpackBackend final
-#ifdef ENABLE_XNNPACK_SHARED_WORKSPACE
-    const std::lock_guard<std::mutex> lock(workspace_mutex_);
-#endif
+    std::unique_lock<std::mutex> lock(workspace_mutex_, std::defer_lock);
```
Yeah. I'm not too worried about this for most use cases, as they'll be fighting for compute anyways, but it would be good to not have that limitation. The issue that I'm actually most concerned about is not reclaiming memory.

I'm going to re-work this diff to use backend options, and maybe we should tackle workspace per model at the same time. I'll take a look at the backend interface and propose a way to identify the method that the partition came from.

digantdesai replied:
> Should we do some more thinking into how workspaces should be managed among different methods?

We discussed it here - https://discord.com/channels/1334270993966825602/1336777807509979188/1375180079679668296

GregoryComer commented:
Reworking this to use the backend options API. It's WIP again. I don't see a way to revert to draft, but I'm validating CI with the initial changes.

GregoryComer commented:
Changes are code complete. I'm waiting on full CI run, but this should be ready for review again.

CC @mcr229 @digantdesai

digantdesai reviewed (commented):
still reviewing but want to drop this before I get back to this

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+          backend_options) override {
+    if (backend_options.size() > 0) {
+      for (const auto& option : backend_options) {
+        if (strcmp(option.key, xnnpack::workspace_sharing_mode_option_key) ==
```
Yeah, I was thinking about introducing an abstraction around backend options, but didn't want to bloat this PR. I can take this a follow-up.

GregoryComer replied:
Filed as https://github.com/pytorch/executorch/issues/13190. I added some brief comments in the issue, but it would be nice create a reusable abstraction that multiple backends can leverage. But I don't know if we can get it in the core runtime (space + embedded constraints), and I don't want to add it as an extension (I don't think it makes sense). Any thoughts?

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+
+            ET_LOG(
+                Debug, "Setting XNNPACK workspace sharing mode to %d.", *val);
+            workspace_sharing_mode_ = static_cast<WorkspaceSharingMode>(*val);
```
The current design associates the mode with the executor (implicitly via the XNNWorkspace object), so it should be safe to set in all scenarios. I believe almost all of these scenarios are covered by the tests that load two models while changing mode in between. I can add a little more coverage there, if desired.

GregoryComer replied:
I've added some extra coverage for running while changing modes.

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.h:
```diff
@@ -0,0 +1,44 @@
+  /// This reduces memory consumption
+  /// for methods with multiple delegate calls, at the cost of only allowing one
+  /// method to execute at a time.
+  PerModel = 1,
```
Yeah, I'd like to do this as a follow-up. I considered adding a per-method (instance) sharing mode (see PR description), but we need a way to identify method instances (as opposed to the name). I think that makes sense as the default behavior going forward, to avoid breaking the use cases that rely on loading the same method multiple times to execute from different threads. Thoughts?

cccclai commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -86,27 +82,30 @@ class XnnpackBackend final
-    const std::lock_guard<std::mutex> lock(workspace_mutex_);
-#endif
+
+    auto program_id =
+        reinterpret_cast<uintptr_t>(context.get_runtime_allocator());
```
> MallocMemoryAllocator across various methods this pointer

Is it supported? I think each method will have their dedicated allocator, otherwise the memory might be corrupted

digantdesai replied:
> this ends up being a larger pain due to thread sychronization in the core

Not sure I understand but i don't want to block you. Please leave a TODO(#ISSUE) comment and I am ok with moving on for now.


> we can make larger PAL changes to allow for generating unique IDs in multithreaded environments

yeah that seems like the right place to generate a unique id for the runtime based on `uuid = fn(pid, tid, pte_id, method_id, more..)`

>  @JacobSzwejbka suggested using signed PTE signatures in the future.

What if a PTE loaded twice by the same process but don't want to share workspace.

>  I think each method will have their dedicated allocator, otherwise the memory might be corrupted

Can you elaborate? IIUC this is just malloc() under the hood. `a->malloc()` twice vs. `a->malloc()` and `b->malloc()`. You can cause the internal vector.emplace to cause race but if you lock at the method load not sure what can be a problem.

cccclai replied:
Are a and b the allocator used by two methods? And `a->malloc` means the allocator used by method A and `b->malloc` means the allocator used by method B?

digantdesai replied:
a, b are the `MallocMemoryAllocator` instance pointers.

digantdesai reviewed (commented):
Some more comments. I will work with you to make sure we can land this today

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+    std::scoped_lock<std::mutex> lock(workspace_meta_mutex_);
+
+    // Check for an existing (live) workspace for this program.
+    auto match = model_workspaces_.find(program_id);
```
Yeah, I don't love this, but it should be safe in all cases. Either you end up using more memory than needed or it enforces extra synchronization. I'd like to push more on the method/program ID.

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+  }
+
+  // Retrieve the global workspace, lazy initializing it if needed.
+  Result<std::shared_ptr<XNNWorkspace>> get_global_workspace() const {
```
Updated.

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+
+    // Lazy init global workspace.
+    if (!global_workspace_) {
+      ET_LOG(Debug, "Instantiating global workspace.");
```
Done.

GregoryComer commented on backends/xnnpack/runtime/XNNPACKBackend.cpp:
```diff
@@ -175,32 +165,179 @@ class XnnpackBackend final
+  Error get_option(
+      BackendOptionContext& context,
+      executorch::runtime::Span<executorch::runtime::BackendOption>&
+          backend_options) override {
```
Updated.

GregoryComer commented on backends/xnnpack/runtime/XNNWorkspace.h:
```diff
@@ -0,0 +1,60 @@
+      return runtime::Error::Internal;
+    }
+
+    return std::make_shared<XNNWorkspace>(
+        WorkspacePtr(workspace, &xnn_release_workspace));
```
Updated.

digantdesai reviewed (approved):
Thanks @GregoryComer

GregoryComer committed:
Add XNNPACK backend option for workspace sharing (#11748)

Summary:
Add a backend option for XNNPACK to enable runtime control of workspace sharing. I've added 3 mode options - Disabled, PerModel, and Global. PerModel shares the workspace between all CALL_DELEGATE instances in a model, keyed by memory allocator address (see below). Global uses a single workspace instance.

I've written the code to allow for the active workspace mode to be safely changed at any time. The workspace instance is resolved at delegate instance init time (model load) and is stored in the XNNExecutor instance. This design will also allow us to set per-model sharing options in the future. I've introduced a wrapper class (XNNWorkspace) to help with synchronization.

With regard to the PerModel behavior, I am using the address of the runtime allocator to disambiguate the model. This is not ideal in the long-run, but there is some larger discussion around generating IDs in a coherent manner in multithreaded environments without synchronization in the core runtime. This might require PAL changes (exposing a thread ID, for example), so I intend to come back to this.

It should be possible to transparently update this logic in the future. The program ID can collide or change without affecting correctness, but may increase memory (for collisions) or enforce extra synchronization (if unstable between delegate instances in a method).

I'd like to add a PerMethod mode as a follow-up. This should be keyed to the specific method instance (not name), such that multiple method instances for the same method can be loaded for execution on different threads without forcing synchronization, but still allow sharing between call delegate instances in each method instance. This will require a unique method identifier.


Test Plan:
CI. I've also added a set of dedicated tests for getting/setting the option, running PTEs in each mode, switching modes at runtime, and I've also updated the multithreaded stress test to run in each mode.

Rollback Plan:

Reviewed By: digantdesai

Differential Revision: D76789804

Pulled By: GregoryComer
